relay-server/src/lib.rs

```diff
@@ -301,9 +303,26 @@ pub fn run(config: Config) -> anyhow::Result<()> {
     // information on all services.
     main_runtime.block_on(async {
         Controller::start(config.shutdown_timeout());
-        let service = ServiceState::start(config.clone())?;
+        let (service, mut join_handles) = ServiceState::start(config.clone())?;
         HttpServer::new(config, service.clone())?.start();
-        Controller::shutdown_handle().finished().await;
+
+        loop {
+            select! {
+                Some(res) = join_handles.next() => {
+                    if let Err(e) = res {
+                        if e.is_panic() {
+                            // Re-trigger panic to terminate the process:
+                            std::panic::resume_unwind(e.into_panic());
+                        }
+                    }
+                }
+                _ = Controller::shutdown_handle().finished() => {
```

Comment:
Note: when every service implements a shutdown listener, awaiting on `finished` becomes unnecessary: We can simply await on all the `join_handles` and guarantee that every service finished its main task.